### PR TITLE
explorer: Add transaction inspector

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
No easy way for devs to share, inspect, and simulate transaction messages. This is often a problem when transactions fail simulation and devs aren't sure why the transaction is failing.

#### Summary of Changes
This adds a new "Transaction Inspector" to the explorer. The inspector provides a convenient ui for decoding transaction messages for debugging purposes. Developer tooling can now emit explorer inspector links with url encoded message data for debugging!

- New inspector route: `/tx/inspector?message=<BASE64 ENCODED MESSAGE>`
- Added simulator card with pretty logging
- Display context for every account (balance, program owner, executable)

##### Empty state:

Screen Shot 2021-06-06 at 1 28 50 AM

##### Transaction inspection:

Screen Shot 2021-06-06 at 12 57 21 AM

##### Pretty logging

Screen Shot 2021-06-06 at 1 24 59 AM

Fixes #

### CODE REVIEW

**Comment by jstarry** [comment]
@mvines fyi, does this seem useful in its current form? I'd like to display account changes in the simulation ui next.

**Comment by mvines** [comment]
> @mvines fyi, does this seem useful in its current form? I'd like to display account changes in the simulation ui next.

This looks really great!  My immediate feedback is that it would be nice to support a copy-paste of a full Transaction in addition to just a Message.  Maybe try to deserialize both?  I feel like a non-zero amount of people will try to paste the full Transaction and then ping for support on Discord when it doesn't work.

**Comment by jon-chuang** [comment]
Would you be able to show some screenshots for when there are errors? Both for system/native program errors, SPL library errors, and user defined program errors.

Currently, on-chain program errors tend to show up as a non human-readable enum value. It would be nice if somehow, custom program errors could return a description of the error?

File: explorer/src/pages/inspector/SimulatorCard.tsx [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+          )
+        );
+
+        // Simulate without signers to skip signer verification
+        const resp = await connection.simulateTransaction(tx);
+
+        let depth = 0;
+        let instructionLogs: InstructionLogs[] = [];
+        const prefixBuilder = (depth: number) => {
+          const prefix = new Array(depth - 1).fill("\u00A0\u00A0").join();
```
Comment (oJshua) [inline explorer/src/pages/inspector/SimulatorCard.tsx:143]:
```suggestion
          const prefix = new Array(depth - 1).fill("\u00A0\u00A0").join("");
```

File: explorer/src/pages/inspector/RawInputCard.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      }
+
+      setRows(Math.max(3, Math.min(10, Math.round(base64.length / 150))));
+
+      let buffer;
+      try {
+        buffer = Uint8Array.from(atob(base64), (c) => c.charCodeAt(0));
+      } catch (err) {
+        console.error(err);
+        setError(err);
```
Comment (oJshua) [inline explorer/src/pages/inspector/RawInputCard.tsx:36]: nit: The exception thrown by atob isn't very user friendly: "Failed to execute 'atob' on 'Window': The string to be decoded is not correctly encoded."

**Review by oJshua — DISMISSED** [review]

**Comment by jstarry** [comment]
> Would you be able to show some screenshots for when there are errors? Both for system/native program errors, SPL library errors, and user defined program errors.
> 
> Currently, on-chain program errors tend to show up as a non human-readable enum value. It would be nice if somehow, transaction inspector could return a description of the error?

@jon-chuang Here's the screenshot. Improving error messages would be great, are you concerned about having standardized error reporting for user deployed programs or the solana native programs?
Screen Shot 2021-06-21 at 3 51 35 PM

File: explorer/src/pages/inspector/SimulatorCard.tsx (hunk 1 above)
Comment (jstarry) [reply explorer/src/pages/inspector/SimulatorCard.tsx:143]: Thanks!

File: explorer/src/pages/inspector/RawInputCard.tsx (hunk 2 above)
Comment (jstarry) [reply explorer/src/pages/inspector/RawInputCard.tsx]: Thanks, changed to `"Input must be base64 encoded"`

**Comment by jstarry** [comment]
> > @mvines fyi, does this seem useful in its current form? I'd like to display account changes in the simulation ui next.
> 
> This looks really great! My immediate feedback is that it would be nice to support a copy-paste of a full Transaction in addition to just a Message. Maybe try to deserialize both? I feel like a non-zero amount of people will try to paste the full Transaction and then ping for support on Discord when it doesn't work.

Full transaction support added!

**Comment by jon-chuang** [comment]
> @jon-chuang Here's the screenshot. Improving error messages would be great, are you concerned about having standardized error reporting for user deployed programs or the solana native programs?
> Screen Shot 2021-06-21 at 3 51 35 PM

Awesome! I think it's probably fine that a non-native program shows up as a `Custom` error as it does on the current explorer. 

Though, I'm not sure if the team is open to maintaining a mapping between the `Custom` enum value and an error string, provided by the bpf program authors. Perhaps, this could be opt-in and on-chain, similar to the way that validators can register some information about themselves like their name and website into an on-chain registry (`solana validator-info`)? Just an idea for future work.

**Comment by jstarry** [comment]
> Though, I'm not sure it might be nice to maintain a mapping between the Custom enum value and an error string, provided by the bpf program authors. Perhaps, this could be opt-in and on-chain, similar to the way that validators can register some information about themselves like their name and website into an on-chain registry (solana validator-info)? Just an idea for future work.

Yeah, I think this should be standardized by the anchor IDL

> For now, native program custom error mapping could possibly be baked in to the explorer via derive(Debug) or otherwise, though. I could work on it in a separate PR if you'd like. This could possibly be handled generically.

Feel free, that would be awesome

**Comment by jon-chuang** [comment]
Ok. As a note to self for the approach: 
Modify snippet below by parsing a `Custom(u32)`, define a different component with the `Pubkey, u32 -> Err string` mapping,  and print as 
`Custom(0x0): YourErrorString`
```
 text: `Program returned error: ${log.slice(
    log.indexOf(": ") + 2
  )}`,
```

**Comment by jstarry** [comment]
> Possibly, we could even include querying error strings as an RPC call... This is trivial for native programs... Not sure how to make this generic over bpf programs... Anchor's solution would rely on a per-program IDL.

Sounds like a good idea, and yeah Anchor supports defining interfaces per program

**Pull request merged**